[Feature][Frontend]:  Continued `stream_options` implementation also in CompletionRequest

Etelis opened the pull request:
This PR introduces support for the `stream_option` parameter in the CompletionRequest class.
Continuing the previous PR: #5135

FIX #4967(*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

Etelis committed:
Add StreamOptions Class

Add stream_options validation in CompletionRequest

Etelis committed:
- Modified the initial response generation to conditionally include `usage` field based on `stream_options.include_usage`.
- Enhanced the token-by-token and finish responses to conditionally include `usage` field if `stream_options.include_usage` is set.
- Added a final usage statistics message if `stream_options.include_usage` is set, including prompt tokens and completion tokens.

Etelis committed:
Tests added for the following scenarios:

- stream=True, stream_options=None

- stream=True, stream_options={"include_usage": True}

- stream=True, stream_options={"include_usage": False}

- stream=False, stream_options={"include_usage": None}

- stream=False, stream_options={"include_usage": False}

- stream=False, stream_options={"include_usage": True}

Etelis committed:
Fixed issues related to formatting.
format.sh.

DarkLight1337 commented:
Unlike Chat Completions API, Completions API doesn't have a response format specific to streaming. Based on the documentation, it seems that `usage` should be returned for each chunk. Can you check whether the real OpenAI API acts in this way?

DarkLight1337 commented on tests/entrypoints/test_openai_server.py:
```diff
@@ -1343,5 +1343,110 @@ async def test_batch_embedding(embedding_server, client: openai.AsyncOpenAI,
+    async for chunk in stream:
+        chunks.append(chunk.choices[0].text)
+    assert len(chunks) > 0
+    assert "usage" not in chunk
```
Should be `chunks` instead of `chunk`

Etelis replied:
Looks like a typo at first, but then I noticed it still wouldn't work, so I changed it to:
`assert all(chunk.usage is None for chunk in chunks)`

Etelis commented:
Will make the needed adjustments later today!
Thank you!

DarkLight1337 commented on tests/entrypoints/test_openai_server.py:
```diff
@@ -1343,5 +1343,110 @@ async def test_batch_embedding(embedding_server, client: openai.AsyncOpenAI,
     assert embeddings.usage.total_tokens == 17
 
 
+async def test_completion_stream_options(server, client: openai.AsyncOpenAI,
```
Should decorate with `@pytest.mark.asyncio`

Etelis commented:
> Unlike Chat Completions API, Completions API doesn't have a response format specific to streaming. Based on the documentation, it seems that `usage` should be returned for each chunk. Can you check whether the real OpenAI API acts in this way?

Actually it seems completions works just the same regarding this:
https://platform.openai.com/docs/api-reference/completions/create

`include_usage

boolean
Optional

If set, an additional chunk will be streamed before the data: [DONE] message. The usage field on this chunk shows the token usage statistics for the entire request, and the choices field will always be an empty array. All other chunks will also include a usage field, but with a null value.`

DarkLight1337 commented:
> > Unlike Chat Completions API, Completions API doesn't have a response format specific to streaming. Based on the documentation, it seems that `usage` should be returned for each chunk. Can you check whether the real OpenAI API acts in this way?
> 
> Actually it seems completions works just the same regarding this: https://platform.openai.com/docs/api-reference/completions/create
> 
> `include_usage
> 
> boolean Optional
> 
> If set, an additional chunk will be streamed before the data: [DONE] message. The usage field on this chunk shows the token usage statistics for the entire request, and the choices field will always be an empty array. All other chunks will also include a usage field, but with a null value.`

Looks like you're right. I focused too much on the return value. In that case, I'll approve the PR once you have fixed the test cases and resolved the merge conflict.

Etelis committed:
FIxing testing file.
Noted by DrakLIght there was two issues:
1. Notation of `@pytest.mark.asyncio` on the test function.
2. Checking on chunk usage on a non exisiting variable.

Etelis committed:
Ran formater.sh to fix formatting issues.

Should have done that on prev commit TBH

Etelis committed:
Merge branch 'main' into Add-support-for-stream_options-in-CompletionRequest

Etelis committed:
Fixed formating.
-- Removed redundent StreamOptions.
-- Formater.sh

DarkLight1337 commented:
I've changed my mind regarding `test_chat_stream_options` - can you fix it in this PR as well?

Etelis committed:
Tests fixture:

-- Added parametrize in completion stream options.

-- Revised streaming tests as the usage is no longer needed to be asserted.

Etelis committed:
Fixing testing.

-- single_usage is no longer needed inside test completion streaming.

DarkLight1337 commented on tests/entrypoints/test_openai_server.py:
```diff
@@ -1342,6 +1340,7 @@ async def test_batch_embedding(embedding_server, client: openai.AsyncOpenAI,
     assert embeddings.usage.total_tokens == 17
 
 
+@pytest.mark.asyncio
```
Need to update this test to actually check Chat Completions API (The code is currently only checking Completions API)

Etelis replied:
Would you prefer these tests to be checked within the same function, or is it better to keep them separate? Currently, I have split them into two functions.
@DarkLight1337

DarkLight1337 replied:
Keep them in separate functions. I'm planning to split up this file in a later PR.

Etelis committed:
Tests Related:

-- Resolved concerns raised by DarkLight (Mistake related to client.chat.completions.creat)

-- Resolved issue related to MODEL_NAME,

Etelis committed:
Fixing Testing.

-- Redundent test removed (stream=True, stream_options=None)

Etelis committed:
Fortmater fixure.

Etelis committed:
Fixed testing.
-- Not found.

DarkLight1337 commented:
Try placing the stream options tests before the embeddings tests. There might be conflicts between the pytest fixtures.

Etelis committed:
Reorder tests in server test file to resolve conflicts

- Moved stream options tests before embeddings tests to address conflicts between pytest fixtures.
- This change is in response to a suggestion from DarkLight1337.
- Adjustments made to ensure test suite runs without errors.

Etelis committed:
Reorder tests in server test file to resolve conflicts

- Moved stream options tests before embeddings tests to address conflicts between pytest fixtures.
- This change is in response to a suggestion from DarkLight1337.
- Adjustments made to ensure test suite runs without errors.
-- Formated code

Etelis committed:
feat(tests): Update test cases for chat and completion streaming options

- **Stream with `include_usage: False`**:
  - Added assertions to ensure no chunk contains the `usage` key.
- **Stream with `include_usage: True`**:
  - Modified test logic to verify that every chunk has `usage` as `None` except for the last chunk, which should have `usage` populated.
- **Stream=False configurations**:
  - Added tests to verify that using `stream_options: {"include_usage": None}`, `{"include_usage": False}`, and `{"include_usage": True}` raises a `BadRequestError`.
- Removed redundant test for `stream=False` with `stream_options: {"include_usage": False}` as it overlaps with the error condition checks.

Etelis committed:
Formatting related issues resolved.

Etelis committed:
fix(tests): Update streaming tests for correct handling of 'usage' attribute

- **Stream with `include_usage: False`**:
  - Updated tests to assert that the `usage` attribute is `None` instead of checking its absence in the chunk dictionary. This aligns with the observed behavior where `usage` is present but set to `None`.

Etelis committed:
Fix: Incorrect indentation causing empty `choices` entries

- Incorrect indent caused an empty `choices` list after each generated `choices`.
- Moved the final `usage` creation step back one indent level to fix this.

Etelis committed:
Fix `serving_chat.py`:

- Incorrect indent caused double sending of a chuck resulting in a server crush.

- Moved the final `usage` creation step back twp indent levels to fix this.

Etelis committed:
Running format.sh.

DarkLight1337 commented:
The entrypoints tests are finally passing. Thanks again for your efforts!

Etelis commented:
> The entrypoints tests are finally passing. Thanks again for your efforts!

Yeah, had some minor issues there. 
But it seems to be fixed now :)

Thank you for the help!!

Etelis committed:
Merge branch 'main' into Add-support-for-stream_options-in-CompletionRequest
